sw/device/silicon_creator/lib/base/sec_mmio.c

```diff
@@ -123,51 +111,39 @@ void sec_mmio_write_increment(uint32_t value) {
 }
 
 void sec_mmio_check_values(uint32_t rnd_offset) {
-  size_t offset = rnd_offset;
+  // Pick a random starting offset.
+  uint32_t offset =
+      ((uint64_t)rnd_offset * (uint64_t)sec_mmio_ctx.last_index) >> 32;
+  uint32_t step = 1;
   size_t i;
-  for (i = 0; i < sec_mmio_ctx.last_index; ++i) {
-    // FIXME: Remove dependency on __udivdi3.
-    offset = (offset + kSecMmioRndStep) % sec_mmio_ctx.last_index;
+  for (i = 0; launder32(i) < sec_mmio_ctx.last_index; ++i) {
     uint32_t read_value = abs_mmio_read32(sec_mmio_ctx.addrs[offset]);
-    if ((read_value ^ kSecMmioMaskVal) != sec_mmio_ctx.values[offset]) {
-      sec_mmio_shutdown_cb(kErrorSecMmioCheckValueFault);
-      __builtin_unreachable();
+    HARDENED_CHECK_EQ((read_value ^ kSecMmioMaskVal),
+                      sec_mmio_ctx.values[offset]);
+    offset += step;
+    if (offset >= sec_mmio_ctx.last_index) {
+      offset -= sec_mmio_ctx.last_index;
     }
   }
   // Check for loop completion.
-  if (i != sec_mmio_ctx.last_index) {
-    sec_mmio_shutdown_cb(kErrorSecMmioCheckIndexFault);
-    __builtin_unreachable();
-  }
+  HARDENED_CHECK_EQ(launder32(i), sec_mmio_ctx.last_index);
```

Comment:
I discussed this with @alphan elsewhere, but we're going to make it part of HARDENED_CHECK's api that it implicitly launders both parameters (which it does: because these are both passed into inline asm constraints, their use is opaque to the compiler, so it cannot learn anything new about them -- just like a launder).

Reply:
SG. Updated the code.